File: Algorithms/4. Median of Two Sorted Arrays.rs
use std::cmp::Ordering;
use std::i32::MAX;

impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let mid = (nums1.len() + nums2.len()) / 2 + 1;
        let mut d1_iter = nums1.iter();
        let mut d2_iter = nums2.iter();
        let mut d1 = d1_iter.next().unwrap_or(&MAX);
        let mut d2 = d2_iter.next().unwrap_or(&MAX);
        
        let mut count = 0;
        let mut result = Vec::<i32>::new();
        loop {
            if (d1 == &MAX && d2 == &MAX) || count == mid {
                break;
            }
            match d1.cmp(&d2) {
                Ordering::Less => {
                    result.push(*d1);
                    d1 = d1_iter.next().unwrap_or(&MAX);
                }
                Ordering::Greater | Ordering::Equal => {
                    result.push(*d2);
                    d2 = d2_iter.next().unwrap_or(&MAX);
                }
            }
            count += 1;
        }
        // let left = mid - 1;
        // let right = mid;
        match (nums1.len() + nums2.len()) % 2 {
            0 => (result[count - 2] + result[count - 1]) as f64 / 2.0,
            _ => result[count - 1] as f64,
        }
    }
}

File: Algorithms/4. Median of Two Sorted Arrays.ts
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
  const m = nums1.length;
  const n = nums2.length;
  const mid = ((m + n) >> 1) + 1;
  let d1 = nums1.shift();
  let d2 = nums2.shift();

  let count = 0;
  const result: number[] = [];
  while (true) {
    if ((d1 === undefined && d2 === undefined) || count === mid) {
      break;
    }
    if (d1 === undefined) {
      result.push(d2);
      d2 = nums2.shift();
      count += 1;
      continue;
    }

    if (d2 === undefined) {
      result.push(d1);
      d1 = nums1.shift();
      count += 1;
      continue;
    }
    if (d1 < d2) {
      result.push(d1);
      d1 = nums1.shift();
    } else {
      result.push(d2);
      d2 = nums2.shift();
    }

    count += 1;
  }
  // console.log(result);
  // console.log((nums1.length + nums2.length));
  // console.log((nums1.length + nums2.length) % 2 === 0);
  return (m + n) % 2 === 0
    ? (result[count - 2] + result[count - 1]) / 2
    : result[count - 1];
}
